### PR TITLE
sync browser page history with ajax updates

### PR DESCRIPTION
**Pull request by braathwaate**
In our FA installation, when a user modifies selection criteria on a FA page and AJAX updates the page, then the user clicks to another page such as via the editor, the browser back button (or FA back link) does not return to the page as the user modified it.  This forces to user to reenter the selection criteria and hit the Show/Search button again.

Perhaps I am missing something, because I can't believe no one else has experienced this issue.

This mod updates the browser page history when AJAX updates the page so that the back button returns to the FA page as the user modified it.

This mod only changes the _7 inquiry_ pages like gl, journal and bank inquiry, although other pages _may_ require this function as well.

### CODE REVIEW

**Comment by apmuthu** [comment]
These changes to `ui_view.inc` and `insert.js` should be carefully integrated as it will affect the inquiry pages in all the extensions as well.

This `ui_view.inc` file has references to functions like `get_js_history()` which are among the newly added functions here which are yet to be included in the core.

**Comment by apmuthu** [comment]
These new functions [stand committed](https://github.com/apmuthu/frontac24/commit/859d01a429bb052d671850ceb9b9870e7aaaf00c) and can be used henceforth.

**Comment by apmuthu** [comment]
The key files of this pull request committed in the core happen to use the splat operator available in PHP 5.6+ only.

[Forum Post](http://frontaccounting.com/punbb/viewtopic.php?pid=31055#p31055)

[Splat Emulation](https://stackoverflow.com/questions/3570496/best-way-to-emulate-ruby-splat-operator-in-php-function-signatures-method-ove)

**Comment by apmuthu** [comment]
This commit needs to be reverted forthwith as the ajax does not work on the additional fields. For those wishing to experiment with it by making it compatible for PHP version prior to v5.6, the attachment will be useful.
[BW_PHP56_splats.zip](https://github.com/apmuthu/frontac24/files/1796671/BW_PHP56_splats.zip)

**Comment by braathwaate** [comment]
I changed the function to use a passed array instead of variable args.  This should make it compatible with PHP prior to v5.6.

**Comment by apmuthu** [comment]
@braathwaate: Thanks.
Your last file: `core/sales/inquiry/sales_orders_view.ph` should be `core/sales/inquiry/sales_orders_view.php`

Remember that some of the inquiry files have had other changes since your last upstream merge and will need to be synched as well.

If the same ajax array is used more than once, better assign it to an array variable and then pass the array in the functions.

What changes are needed to make the other files have their select dropdown boxes ajax compliant?

File: core/purchasing/inquiry/supplier_allocation_inquiry.php [hunk 1]
```diff
@@ -20,20 +20,10 @@
 	$js .= get_js_open_window(900, 500);
 if (user_use_date_picker())
 	$js .= get_js_date_picker();
+$js .= get_js_history(array('supplier_id', 'TransAfterDate', 'TransToDate'));
```
Comment (apmuthu) [inline core/purchasing/inquiry/supplier_allocation_inquiry.php]: Should be:
`$js .= get_js_history(array('supplier_id', 'FromDate', 'ToDate'));`

File: core/purchasing/inquiry/supplier_allocation_inquiry.php [hunk 2]
```diff
@@ -20,20 +20,10 @@
...
-}
-if (isset($_GET['FromDate']))
-{
-	$_POST['TransAfterDate'] = $_GET['FromDate'];
-}
-if (isset($_GET['ToDate']))
-{
-	$_POST['TransToDate'] = $_GET['ToDate'];
-}
+set_posts(array('supplier_id', 'TransAfterDate', 'TransToDate'));
```
Comment (apmuthu) [inline core/purchasing/inquiry/supplier_allocation_inquiry.php]: Should be:
`set_posts(array('supplier_id', 'FromDate', 'ToDate'));`

File: core/purchasing/inquiry/supplier_allocation_inquiry.php (hunk 1 above)
Comment (apmuthu) [inline core/purchasing/inquiry/supplier_allocation_inquiry.php:23]: Note: The `FromDate` and `ToDate` form fields will now get renamed to `TransAfterDate` and `TransToDate` in line with useage in the codebase and will need to be taken care in the extensions and the RestAPIs as well in all files where such a change occurs.

Add the `filterType` as well but keep the `showSettled` not persistent.

File: core/sales/inquiry/customer_allocation_inquiry.php [hunk 3]
```diff
@@ -22,12 +22,11 @@
...
 	$js .= get_js_date_picker();
+$js .= get_js_history(array("customer_id", "TransAfterDate", "TransToDate", "filterType"));
+
 page(_($help_context = "Customer Allocation Inquiry"), false, false, "", $js);
 
-if (isset($_GET['customer_id']))
-{
-	$_POST['customer_id'] = $_GET['customer_id'];
-}
+set_posts(array('customer_id'));
```
Comment (apmuthu) [inline core/sales/inquiry/customer_allocation_inquiry.php:29]: Why not all fields in the js history here?

**Review by apmuthu — COMMENTED** [review]
After synching with the upstream, see if the attached files satisfy your proposed3 branch.
[Ajax_Post_Retention.zip](https://github.com/apmuthu/frontac24/files/1799666/Ajax_Post_Retention.zip)

http://frontaccounting.com/punbb/viewtopic.php?pid=31085#p31085

File: core/purchasing/inquiry/supplier_allocation_inquiry.php (hunk 1 above)
Comment (braathwaate) [reply core/purchasing/inquiry/supplier_allocation_inquiry.php:23]: Yes, it appears like FromDate and ToDate were never set or used, so renaming them should be fine.

Note that the fields added to set_posts() actually improve the ability of FA to be called from another application using url query strings to display filtered data without requiring user input.

One other note.   FA does call some scripts (like supplier_inquiry.php from po entry to display the credit total) with an url query string.    When called this way, the script suppresses the FA headers.  But headers are shown in the ajax/browser sync.

So the lazy hack I made to discern the two different cases was to check if one of the newly added fields was missing.   This really should be solved instead by passing a flag to suppress the header.    To do this, one would have to identify all the calling scripts and set the suppress_header flag.

File: core/sales/inquiry/customer_allocation_inquiry.php (hunk 3 above)
Comment (braathwaate) [reply core/sales/inquiry/customer_allocation_inquiry.php:29]: Good catch.

**Pull request closed without merging**